src/pip/_internal/commands/install.py

```diff
@@ -416,6 +418,10 @@ def run(self, options: Values, args: List[str]) -> int:
                 # we're not modifying it.
                 modifying_pip = pip_req.satisfied_by is None
             protect_pip_from_modification_on_windows(modifying_pip=modifying_pip)
+            if modifying_pip:
+                # Parallelization will re-import pip when starting new workers
+                # during installation which is unsafe if pip is being modified.
+                options.install_jobs = 1
 
```

Comment:
Since the concurrent executors may spin up new workers midway through package installation, a newly installed pip may be imported which will probably result in a crash ([as seen in CI](https://github.com/pypa/pip/actions/runs/13556467716/job/37891691885?pr=13247#step:7:1037)). This is a shame as parallelization does improve self-install time by ~200ms locally, but alas, pip should not crash even if we're trying to install 10 year old pip.

Technically, we could avoid this by patching `__main__` to point a module that's completely outside of the pip namespace (it has to be outside because it's the import of `pip.__init__` that's blowing up the tests) but I'm strongly opposed to adding a new top-level module such as `__pip_empty.py` to our distributions.

Reply:
To avoid the global impact, we could special-case pip and use a serial compiler no matter what for pip, but parallelization is generally unsafe when modifying pip so this isn't a good idea either.

Reply:
Oh wait, this is even more problematic than I initially thought because the workers could import a new pip was maliciously overwritten. We had a security report similar to this for the self- version check. Hmm, I'll need to go back to the drawing board. Hopefully there is a way to avoid the extra imports on worker startup, otherwise, I'm going to have to use multiprocessing.Pool directly and pay the worker startup overhead all at once... :(